test/python/model_validation/validation_tool.py

```diff
@@ -0,0 +1,184 @@
+#             neg_log_likelihood = outputs.loss
+
+#         nlls.append(neg_log_likelihood)
+
+#         prev_end_loc = end_loc
+#         if end_loc == seq_len:
+#             break
+
+#     ppl = torch.exp(torch.stack(nlls).mean())
+#     return
+
+
+
+def validate_model(args, output_dir):
+    if args.verbose: print("Loading model...")
+    if args.timings:
+        started_timestamp = 0
+        first_token_timestamp = 0
+
+    model = og.Model(f'{output_dir}')
+    if args.verbose: print("Model loaded")
+    tokenizer = og.Tokenizer(model)
+    tokenizer_stream = tokenizer.create_stream()
+    if args.verbose: print("Tokenizer created")
+    if args.verbose: print()
+    search_options = {name:getattr(args, name) for name in ['do_sample', 'max_length', 'min_length', 'top_p', 'top_k', 'temperature', 'repetition_penalty'] if name in args}
+    
+    # Set the max length to something sensible by default, unless it is specified by the user,
+    # since otherwise it will be set to the entire context length
+
+    search_options['max_length'] = 512
+
+    chat_template = '<|user|>\n{input} <|end|>\n<|assistant|>'
+
+
+    inputs = ["Provide a detailed analysis of the causes and consequences of the French Revolution, including key events, figures, and social changes.",
+              "Explain the process of photosynthesis in plants, detailing the chemical reactions involved, the role of chlorophyll, and the importance of sunlight"]   
+    
+    for input in inputs:
+
+        complete_text = ''
+        
+        if args.timings: started_timestamp = time.time()
+
+        # If there is a chat template, use it
+        prompt = f'{chat_template.format(input=input)}'
+
+
+        # Tokenizer has interesting behavior, it creates multiple inputs
+        input_tokens = tokenizer.encode(prompt)
+
+
+        params = og.GeneratorParams(model)
+        params.set_search_options(**search_options)
+        params.input_ids = input_tokens
+
+        generator = og.Generator(model, params)
+        if args.verbose: print("Generator created")
+
+        if args.verbose: print("Running generation loop ...")
+        if args.timings:
+            first = True
+            new_tokens = []
+
+        print()
+        print("Output: ", end='', flush=True)
+
+        try:
+            while not generator.is_done():
+                generator.compute_logits()
+                generator.generate_next_token()
+                if args.timings:
+                    if first:
+                        first_token_timestamp = time.time()
+                        first = False
+
+                new_token = generator.get_next_tokens()[0]
+                
+                value_to_save = tokenizer_stream.decode(new_token)
+
+                complete_text += value_to_save
+
+                print(tokenizer_stream.decode(new_token), end='', flush=True)
+
+                
+                if args.timings: new_tokens.append(new_token)
+        except KeyboardInterrupt:
```

Comment:
>except KeyboardInterrupt: [](http://example.com/codeflow?start=8&length=25)

if other exceptions are thrown, catch the exception, log, and the validation for current model failed.